### PR TITLE
Mid: storage-mon: daemonize storage_mon to deal with I/O hangs.(Add daemon/client mode)

### PR DESCRIPTION
**Pull request by HideoYamauchi**
Hi All,

This fix is a modified version of the following PR.
 - https://github.com/ClusterLabs/resource-agents/pull/1862

This PR reflects the comments in the previous PR.

Please comment on this fix.

Best Regards,
Hideo Yamauchi.

### CODE REVIEW

File: heartbeat/storage-mon.in [hunk 1]
```diff
@@ -48,20 +48,26 @@
...
+ATTRDUP=${HA_SBIN_DIR}/attrd_updater
+PIDFILE=${HA_VARRUN}/storage-mon-${OCF_RESOURCE_INSTANCE}.pid
+ATTRNAME="#health-${OCF_RESOURCE_INSTANCE}"
 
 OCF_RESKEY_CRM_meta_interval_default="0"
 OCF_RESKEY_io_timeout_default="10"
+OCF_RESKEY_check_interval_default="30"
 OCF_RESKEY_inject_errors_default=""
 OCF_RESKEY_state_file_default="${HA_RSCTMP%%/}/storage-mon-${OCF_RESOURCE_INSTANCE}.state"
+OCF_RESKEY_daemonize_default=""
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:61]: You should set this to false, so we can use ocf_is_true later.

File: heartbeat/storage-mon.in [hunk 2]
```diff
@@ -114,6 +128,14 @@ Used only for testing! Specify % of I/O errors to simulate drives failures.
 <content type="integer" default="${OCF_RESKEY_inject_errors_default}" />
 </parameter>
 
+<parameter name="daemonize" unique="0">
+<longdesc lang="en">
+Specifies to start storage-mon as a daemon and check for devices.
+</longdesc>
+<shortdesc lang="en">start storage-mon with daemon</shortdesc>
+<content type="string" default="" />
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:136]: Should be `default="${OCF_RESKEY_daemonize_default}"`.

File: heartbeat/storage-mon.in [hunk 3]
```diff
@@ -173,77 +205,182 @@ storage-mon_init() {
...
+			# If the attrd_updater fails with the red attribute, return an error to let pacemaker handle the failure immediately.
+			return $OCF_ERR_GENERIC
+		fi
+	done
 	return $OCF_SUCCESS
 }
 
 storage-mon_monitor() {
-	storage-mon_init
+	if [ -z "$OCF_RESKEY_daemonize" ]; then
```
Comment (oalbrigt) [inline heartbeat/storage-mon.in:228]: Should be `if ! ocf_is_true "$OCF_RESKEY_daemonize"; then`. Update other occurences as well.

**Review by oalbrigt — CHANGES_REQUESTED** [review]
I've added a couple of comments for improvements, and you have to add `BuildRequires: libqb-devel` and `Requires: libqb` to the spec-file.

**Comment by wenningerk** [comment]
>add `BuildRequires: libqb-devel` and `Requires: libqb` to the spec-file.

Isn't the BuildRequires-Part already in the patch? And shouldn't the Requires-Part be automatically generated?

**Comment by oalbrigt** [comment]
> > add `BuildRequires: libqb-devel` and `Requires: libqb` to the spec-file.
> 
> Isn't the BuildRequires-Part already in the patch? And shouldn't the Requires-Part be automatically generated?

Ah. I see. I was trying to build on one of the distros, and forgot to diff the spec :)

The Requires part isnt auto generated though. But will probably be present be present in most cases due to Pacemaker or deps depending on it. So adding it is only to ensure it's working if anyone is installing it without those.

**Comment by oalbrigt** [comment]
> > > add `BuildRequires: libqb-devel` and `Requires: libqb` to the spec-file.
> > 
> > 
> > Isn't the BuildRequires-Part already in the patch? And shouldn't the Requires-Part be automatically generated?
> 
> Ah. I see. I was trying to build on one of the distros, and forgot to diff the spec :)
> 
> The Requires part isnt auto generated though. But will probably be present be present in most cases due to Pacemaker or deps depending on it. So adding it is only to ensure it's working if anyone is installing it without those.

Correct. The Requires is auto-generated as well, so ignore the spec-part of my review :)

**Comment by HideoYamauchi** [comment]
Hi Oyvind,

Check and correct the comments for OCF_RESKEY_daemonize_default and OCF_RESKEY_daemonize.

Many thanks,
Hideo Yamauchi.

**Comment by oalbrigt** [comment]
@HideoYamauchi They are correct. We should use ocf_is_true 

> Hi Oyvind,
> 
> Check and correct the comments for OCF_RESKEY_daemonize_default and OCF_RESKEY_daemonize.
> 
> Many thanks, Hideo Yamauchi.

Done.

**Review by oalbrigt — APPROVED** [review]

**Comment by oalbrigt** [comment]
Thanks.

**Comment by HideoYamauchi** [comment]
Hi Oyvind,

Thanks for merging this fix.

Best Regards,
Hideo Yamauchi.

**Pull request merged**